roomserver/internal/perform_join.go

```diff
@@ -204,11 +204,12 @@ func (r *RoomserverInternalAPI) performJoinRoomByID(
 		// a member of the room.
 		alreadyJoined := false
 		for _, se := range buildRes.StateEvents {
+			if !se.StateKeyEquals(*event.StateKey()) {
```

Comment:
We probably want to preserve `se.StateKey() != nil` here for safety. I don't *think* there's a reason that the query API would return something strange to the builder but it's not worth panicking if it does.

Reply:
This nil-check is already implemented inside `StateKeyEquals`

Reply:
It is for `se`, but not for `*event` which has to dereference into a `string` type or it'll likely panic.

Reply:
I can simply replace  `*event.StateKey()` with `userID`

Reply:
Perhaps the sensible thing to do is `event.StateKeyEquals(userID)` after the event is built. Then you should be able to be confident about `*event.StateKey()` being OK on line 207.

Reply:
I feel it would be like testing that the implementation of the builder is right.

Maybe I could add a simple unit-test for that (calling the builder just like in this function and checking the `StateKeyEquals`).

Then I can use either `se.StateKeyEquals(userID)` or `se.StateKeyEquals(*event.StateKey())` without having to think about it :-)